diff --git a/src/nix-env/nix-env.cc b/src/nix-env/nix-env.cc
index e65dc4e..8e70859 100644
--- a/src/nix-env/nix-env.cc
+++ b/src/nix-env/nix-env.cc
@@ -4,6 +4,7 @@
 #include "misc.hh"
 #include "shared.hh"
 #include "eval.hh"
+#include "eval-inline.hh"
 #include "get-drvs.hh"
 #include "attr-path.hh"
 #include "common-opts.hh"
@@ -1139,6 +1140,88 @@ static void opQuery(Globals & globals, Strings opFlags, Strings opArgs)
                                 }
                             }
                         }
+
+                        const Symbol  sRepositories(globals.state->symbols.create("repositories")),
+                                      sGit(globals.state->symbols.create("git")),
+                                      sSrc(globals.state->symbols.create("src")),
+                                      sRev(globals.state->symbols.create("rev")),
+                                      sUrl(globals.state->symbols.create("url")),
+                                      sUrls(globals.state->symbols.create("urls")),
+                                      sOutputHash(globals.state->symbols.create("outputHash"));
+
+                        // meta.repositories special case handling
+                        Bindings::iterator meta_raw = i.attrs->find(globals.state->sMeta);
+                        if (meta_raw != i.attrs->end()) {
+                            globals.state->forceAttrs(*meta_raw->value);
+                            Bindings::iterator repositories = meta_raw->value->attrs->find(sRepositories);
+                            if (repositories != meta_raw->value->attrs->end()) {
+                                globals.state->forceAttrs(*repositories->value);
+                                Bindings::iterator git = repositories->value->attrs->find(sGit);
+                                if (git != repositories->value->attrs->end()) {
+                                    globals.state->forceValue(*git->value);
+                                    if (git->value->type == tString) {
+                                        XMLAttrs attrs4;
+                                        attrs4["name"] = "repositories.git";
+                                        attrs4["type"] = "string";
+                                        attrs4["value"] = git->value->string.s;
+                                        xml.writeEmptyElement("meta", attrs4);
+                                    }
+                                }
+                            }
+                        }
+
+
+                        // src attr special case handling
+                        Bindings::iterator src = i.attrs->find(sSrc);
+                        if (src != i.attrs->end()) {
+                            globals.state->forceValue(*src->value);
+                            if (src->value->type == tAttrs) {
+                                Bindings::iterator urls = src->value->attrs->find(sUrls);
+                                if (urls != src->value->attrs->end()) {
+                                    globals.state->forceList(*urls->value);
+                                    if (urls->value->listSize() > 0) {
+                                        XMLAttrs attrs5;
+                                        Value *url = urls->value->listElems()[0];
+
+                                        attrs5["name"] = "src.url";
+                                        attrs5["type"] = "string";
+                                        attrs5["value"] = globals.state->forceStringNoCtx(*url);
+                                        xml.writeEmptyElement("meta", attrs5);
+                                    }
+                                }
+                                Bindings::iterator url = src->value->attrs->find(sUrl);
+                                if (url != src->value->attrs->end()) {
+                                   XMLAttrs attrs6;
+                                   attrs6["name"] = "src.repo";
+                                   attrs6["type"] = "string";
+                                   attrs6["value"] = globals.state->forceStringNoCtx(*url->value);
+                                   xml.writeEmptyElement("meta", attrs6);
+                                }
+                                Bindings::iterator rev = src->value->attrs->find(sRev);
+                                if (rev != src->value->attrs->end()) {
+                                   globals.state->forceValue(*rev->value);
+                                   XMLAttrs attrs7;
+                                   attrs7["name"] = "src.rev";
+                                   attrs7["type"] = "string";
+                                   if (rev->value->type == tString) {
+                                       attrs7["value"] = rev->value->string.s;
+                                   } else if (rev->value->type == tInt) {
+                                       attrs7["value"] = (format("%1%") % rev->value->integer).str();
+                                   }
+                                   xml.writeEmptyElement("meta", attrs7);
+                                }
+                                Bindings::iterator sha256 = src->value->attrs->find(sOutputHash);
+                                if (sha256 != src->value->attrs->end()) {
+                                   XMLAttrs attrs6;
+                                   attrs6["name"] = "src.sha256";
+                                   attrs6["type"] = "string";
+                                   attrs6["value"] = globals.state->forceStringNoCtx(*sha256->value);
+                                   xml.writeEmptyElement("meta", attrs6);
+                                }
+                            }
+                        }
+
+
                     }
                 } else
                     xml.writeEmptyElement("item", attrs);
